src/ansi-c/library/cprover_contracts.c

```diff
@@ -1191,17 +1192,40 @@ __CPROVER_HIDE:;
         (write_set->assert_ensures_ctx == 0),
       "only one context flag at a time");
 #endif
-    // fail if size is too big
-    if(size >= __CPROVER_max_malloc_size)
-      return 0;
+    if(
+      __CPROVER_malloc_failure_mode ==
+      __CPROVER_malloc_failure_mode_return_null)
+    {
+      // When --malloc-may-fail --malloc-fail-null
+      // add implicit assumption that the size is capped
+      __CPROVER_assume(size < __CPROVER_max_malloc_size);
+    }
+    else if(
+      __CPROVER_malloc_failure_mode ==
+      __CPROVER_malloc_failure_mode_assert_then_assume)
+    {
+      // When --malloc-may-fail --malloc-fail-assert
+      // check if max allocation size is exceeded and
+      // add implicit assumption that the size is capped
+      __CPROVER_assert(
+        size < __CPROVER_max_malloc_size,
+        "__CPROVER_is_fresh max allocation size exceeded");
+      __CPROVER_assume(size < __CPROVER_max_malloc_size);
+    }
 
-    // pass a null write set pointer to the instrumented malloc
-    void *ptr = __CPROVER_contracts_malloc(size, 0);
+    void *ptr = __CPROVER_allocate(size, 0);
     *elem = ptr;
-    // malloc can also return a NULL pointer if failure modes are active
-    if(!ptr)
-      return 0;
-      // record fresh object in the object set
+
+    // record the object size for non-determistic bounds checking
+    __CPROVER_bool record_malloc = __VERIFIER_nondet___CPROVER_bool();
+    __CPROVER_malloc_is_new_array =
+      record_malloc ? 0 : __CPROVER_malloc_is_new_array;
+
+    // detect memory leaks
+    __CPROVER_bool record_may_leak = __VERIFIER_nondet___CPROVER_bool();
+    __CPROVER_memory_leak = record_may_leak ? ptr : __CPROVER_memory_leak;
```

Comment:
Should uses of `is_fresh` really participate in memory-leak checking?

Reply:
`is_fresh` allocates when assumed
1.  in a requires clause of a contract being checked;
  - should not participate in leak checking because case 1. models what's provided by the open environmnent of the contract being checked.
2. in an ensures clause of a contract being used to abstract a call.
  - should participate in leak checking because it models an allocation happening in the user program, not the environment

Does that seem reasonable ?

Reply:
Sounds good to me.